### PR TITLE
VideoBufferDMA: Support exporting YCbCr444 buffers

### PR DESCRIPTION
**Pull request by popcornmix**
## Description
Fix incorrect colours with YCbCr444 pixel format and EGL rendering

## Motivation and context
The current code assumes chroma is decimated by two, but that is not necessarily the case. DRMPRIME decode with EGL rendering of YCbCr444 video will have corrupted colours.

Ask ffmpeg what the chroma decimation is.

## How has this been tested?
Tested on Pi4 using DRMPRIME decoder and EGL rendering

[This](http://v2v.cc/~j/theora_testsuite/ducks_take_off_444_720p25.ogg) is a suitable test file

## What is the effect on users?
No more incorrect colours

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by popcornmix** [comment]
Pulled the initial offset calculations out of loop to keep the inside simpler.

File: xbmc/cores/VideoPlayer/Buffers/VideoBufferDMA.cpp [hunk 1]
```diff
@@ -119,20 +119,29 @@ bool CVideoBufferDMA::Alloc()
 
 void CVideoBufferDMA::Export(AVFrame* frame, uint32_t width, uint32_t height)
 {
-  m_planes = av_pix_fmt_count_planes(static_cast<AVPixelFormat>(frame->format));
+  AVPixelFormat pix_fmt = static_cast<AVPixelFormat>(frame->format);
+  m_planes = av_pix_fmt_count_planes(pix_fmt);
+  int h_shift, v_shift;
+
+  if (av_pix_fmt_get_chroma_sub_sample(pix_fmt, &h_shift, &v_shift))
+    throw std::runtime_error(
```
Comment (garbear) [inline xbmc/cores/VideoPlayer/Buffers/VideoBufferDMA.cpp:127]: Is a failure in an external library really something we want to hard crash to desktop for?

File: xbmc/cores/VideoPlayer/Buffers/VideoBufferDMA.cpp [hunk 2]
```diff
@@ -119,20 +119,29 @@ bool CVideoBufferDMA::Alloc()
...
 
   for (uint32_t plane = 0; plane < m_planes; plane++)
-  {
     m_strides[plane] =
         av_image_get_linesize(static_cast<AVPixelFormat>(frame->format), width, plane);
+
+  m_offsets[0] = 0;
+  m_offsets[1] = m_strides[0] * height;
+  for (uint32_t plane = 2; plane < m_planes; plane++)
     m_offsets[plane] =
```
Comment (lrusak) [inline xbmc/cores/VideoPlayer/Buffers/VideoBufferDMA.cpp:143]: Off by one error I think?

Comment (popcornmix) [reply xbmc/cores/VideoPlayer/Buffers/VideoBufferDMA.cpp:143]: Could you explain exactly what you think the correction is? It's looking okay to me.
plane(N) = offset_of_plane(N-1) + size_of_plane(N-1).
And we've already done plane(0) and plane(1).

Comment (lrusak) [reply xbmc/cores/VideoPlayer/Buffers/VideoBufferDMA.cpp:143]: Yea I ran some iterations and it seems fine. I don't think we need the loop at all though and we can just be explicit about what is happening
```c++
  m_offsets[0] = 0;
  m_offsets[1] = m_offsets[0] + (m_strides[0] * (height >> 0));
  m_offsets[2] = m_offsets[1] + (m_strides[1] * (height >> v_shift));
```

There may not always be three planes but we allocate 3 always.

Comment (popcornmix) [reply xbmc/cores/VideoPlayer/Buffers/VideoBufferDMA.cpp]: In theory there could be formats with an additional plane (e.g. alpha), so I left the loop in, but as you say it is initialised to 3, and handling a format with extra planes may well require extra logic so I've removed it.

I've made sanity check of `m_planes >= 2` also require `<=3`.

File: xbmc/cores/VideoPlayer/Buffers/VideoBufferDMA.cpp [hunk 3]
```diff
@@ -119,20 +119,25 @@ bool CVideoBufferDMA::Alloc()
 
 void CVideoBufferDMA::Export(AVFrame* frame, uint32_t width, uint32_t height)
 {
-  m_planes = av_pix_fmt_count_planes(static_cast<AVPixelFormat>(frame->format));
+  AVPixelFormat pix_fmt = static_cast<AVPixelFormat>(frame->format);
+  m_planes = av_pix_fmt_count_planes(pix_fmt);
+  int h_shift, v_shift;
```
Comment (lrusak) [inline xbmc/cores/VideoPlayer/Buffers/VideoBufferDMA.cpp:124]: Please declare each variable on a single line.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#51-multiple-declarations

File: xbmc/cores/VideoPlayer/Buffers/VideoBufferDMA.cpp [hunk 4]
```diff
@@ -119,20 +119,25 @@ bool CVideoBufferDMA::Alloc()
...
+  if (av_pix_fmt_get_chroma_sub_sample(pix_fmt, &h_shift, &v_shift))
+    throw std::runtime_error("unable to determine chroma_sub_sample: " +
+                             std::string(av_get_pix_fmt_name(pix_fmt)));
+
+  if (m_planes < 2 || m_planes > 3)
+    throw std::runtime_error("only 2 or 3 plane formats supported: " +
+                             std::string(av_get_pix_fmt_name(pix_fmt)));
 
   for (uint32_t plane = 0; plane < m_planes; plane++)
-  {
```
Comment (lrusak) [inline xbmc/cores/VideoPlayer/Buffers/VideoBufferDMA.cpp:130]: I'd rather keep these braces for clarity please.

Comment (popcornmix) [reply xbmc/cores/VideoPlayer/Buffers/VideoBufferDMA.cpp:130]: Updated

**Review by lrusak — APPROVED** [review]
Looks good, thanks for the improvement!

**Pull request merged**